File: java/src/kanzi/io/BlockEvent.java
/*
Copyright 2011-2013 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.io;

public class BlockEvent 
{
   public enum Type
   {
      BEFORE_TRANSFORM,
      AFTER_TRANSFORM,
      BEFORE_ENTROPY,
      AFTER_ENTROPY
   }

   private final int id;
   private final int size;
   private final int hash;
   private final Type type;
   private final boolean hashing;
   private final long time;
   private final long unixTime;
   
   
   public BlockEvent(Type type, int id, int size)
   {
      this(type, id, size, 0, false);
   }

   
   public BlockEvent(Type type, int id, int size, int hash)
   {
      this(type, id, size, hash, true);
   }
   
   
   protected BlockEvent(Type type, int id, int size, int hash, boolean hashing)
   {
      this.id = id;
      this.size = size;
      this.hash = hash;
      this.hashing = hashing;
      this.type = type;
      this.time = System.nanoTime();
      this.unixTime = System.currentTimeMillis();
   }
   
   
   public int getId() 
   {
      return this.id;
   }

   
   public int getSize() 
   {
      return this.size;
   }

   
   public long getTime() 
   {
      return this.time;
   }

   
   public Integer getHash() 
   {
      return (this.hashing == false) ? null : this.hash;
   }  
   
   
   public Type getType()
   {
      return this.type;
   }


   public long getUnixTime()
   {
      return this.unixTime;
   }
   
   
   @Override
   public String toString()
   {
      StringBuilder sb = new StringBuilder(200);
      sb.append("{ \"type\":\"").append(this.getType()).append("\"");
      sb.append(", \"id\":").append(this.getId());
      sb.append(", \"size\":").append(this.getSize());
      sb.append(", \"time\":").append(this.getUnixTime());
      
      if (this.hashing == true)
         sb.append(", \"hash\":").append(Integer.toHexString(this.getHash()));

      sb.append(" }");
      return sb.toString();
   }
}

File: go/src/kanzi/io/BlockEvent.go
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package io

import (
	"fmt"
	"time"
)

const (
	EVT_BEFORE_TRANSFORM = 0
	EVT_AFTER_TRANSFORM  = 1
	EVT_BEFORE_ENTROPY   = 2
	EVT_AFTER_ENTROPY    = 3
)

type BlockEvent struct {
	eventType int
	blockId   int
	blockSize int
	hash      uint32
	hashing   bool
	time_     time.Time
}

func (this *BlockEvent) EventType() int {
	return this.eventType
}

func (this *BlockEvent) BlockId() int {
	return this.blockId
}

func (this *BlockEvent) Time() time.Time {
	return this.time_
}

func (this *BlockEvent) BlockSize() int {
	return this.blockSize
}

func (this *BlockEvent) Hash() uint32 {
	return this.hash
}

func (this *BlockEvent) Hashing() bool {
	return this.hashing
}

func (this *BlockEvent) String() string {
	hash := ""
	type_ := ""

	if this.hashing == true {
		hash = fmt.Sprintf(", \"hash\": %x", this.hash)
	}

	if this.eventType == EVT_BEFORE_TRANSFORM {
		type_ = "BEFORE_TRANSFORM"
	} else if this.eventType == EVT_AFTER_TRANSFORM {
		type_ = "AFTER_TRANSFORM"
	} else if this.eventType == EVT_BEFORE_ENTROPY {
		type_ = "BEFORE_ENTROPY"
	} else if this.eventType == EVT_AFTER_ENTROPY {
		type_ = "AFTER_ENTROPY"
	}

	return fmt.Sprintf("{ \"type\":\"%s\", \"id\":%d, \"size\":%d, \"time\":%d%s}", type_, this.blockId, this.blockSize,
		this.time_.UnixNano()/1000000, hash)
}

type BlockListener interface {
	ProcessEvent(evt *BlockEvent)
}
